server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -608,78 +655,128 @@ public static boolean isDedicatedVotingOnlyNode(Set<DiscoveryNodeRole> roles) {
         try {
             final ClusterState state = event.state();
-            assert assertReadonlyRepositoriesNotInUseForWrites(state);
-            final RepositoriesMetadata oldMetadata = RepositoriesMetadata.get(event.previousState());
-            final RepositoriesMetadata newMetadata = RepositoriesMetadata.get(state);
-
-            // Check if repositories got changed
-            if (oldMetadata.equalsIgnoreGenerations(newMetadata)) {
-                for (Repository repo : repositories.values()) {
-                    repo.updateState(state);
-                }
-                return;
+            final ClusterState previousState = event.previousState();
+
+            for (var projectId : event.projectDelta().removed()) { // removed projects
+                applyProjectState(state.version(), null, previousState.projectState(projectId));
             }
 
-            logger.trace("processing new index repositories for state version [{}]", event.state().version());
+            for (var projectId : event.projectDelta().added()) { // added projects
+                applyProjectState(state.version(), state.projectState(projectId), null);
+            }
 
-            Map<String, Repository> survivors = new HashMap<>();
-            // First, remove repositories that are no longer there
-            for (Map.Entry<String, Repository> entry : repositories.entrySet()) {
-                if (newMetadata.repository(entry.getKey()) == null) {
-                    logger.debug("unregistering repository [{}]", entry.getKey());
-                    Repository repository = entry.getValue();
-                    closeRepository(repository);
-                    archiveRepositoryStats(repository, state.version());
-                } else {
-                    survivors.put(entry.getKey(), entry.getValue());
-                }
+            for (var projectId : event.projectDelta().common()) { // existing projects
+                applyProjectState(state.version(), state.projectState(projectId), previousState.projectState(projectId));
+            }
+        } catch (Exception ex) {
+            assert false : new AssertionError(ex);
+            logger.warn("failure updating cluster state ", ex);
+        }
+    }
+
+    /**
+     * Apply changes for one project. The project can be either newly added, removed or an existing one.
+     *
+     * @param version The cluster state version of the change.
+     * @param state The current project state, or {@code null} if the project was removed.
+     * @param previousState the previous project state, or {@code null} if the project was newly added.
+     */
+    private void applyProjectState(long version, @Nullable ProjectState state, @Nullable ProjectState previousState) {
+        assert state != null || previousState != null : "state and previousState cannot both be null";
+        assert state == null || assertReadonlyRepositoriesNotInUseForWrites(state);
```

Comment:
The changeset of this method can be a bit deceptively large due to method extraction and indentation difference. It is largely the same as the exisitng `applyClusterState` method with some additional logic for new and removed projects.

Reply:
nit: Any chance we could break up this method so not both states can be nullable so it is a bit more readable? I find it rather odd but I understand that's because you want to keep this logic the same with just some null checks.

Reply:
I initially started with breaking up the method for added, removed and existing projects. But found out that processing existing projects requires all logic (because repo can still be added/removed for existing projects) which led to duplication and felt non-ideal. That said, I tried it again by separating "removed projects" out in its own method and it seems to be a better middle-ground (compared to 1 or 3 methods). Please let me know whether it looks better to you. f62e3fe169c6a75823b3b3a100510a0c5d0ac781